File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        
        int n=height.length;
        int leftmax=0;
        int rightmax=0;
        int left=0;
        int right=n-1;

        int area=0;
        while(left<=right){
            if(leftmax<rightmax){
                leftmax=Math.max(leftmax, height[left]); //to avoaid writing area>=0, we are updating left max
                area+= leftmax-height[left];
                left++;
            }
            else{
                 rightmax=Math.max(rightmax, height[right]);
                 area+= rightmax- height[right];
                 right--;
            }
        }
        return area;
    }
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        
        res=0
        leftmax=0
        rightmax=0
        l=0
        r=len(height)-1

        while l<=r:
            leftmax= max(height[l],leftmax)
            rightmax=max(height[r],rightmax)
            if leftmax<rightmax:
                res+= leftmax-height[l]
                l+=1
            else:
                res+=rightmax-height[r]
                r-=1
            
        return res